runtime/runc_jailer.go

```diff
@@ -0,0 +1,399 @@
+				return err
+			}
+			m.Cfg.LogFifo = newFifoPath
+
+			metricFifoFileName := filepath.Base(m.Cfg.MetricsFifo)
+			newMetricFifoPath := filepath.Join(contentsPath, metricFifoFileName)
+			if err := os.Link(m.Cfg.MetricsFifo, newMetricFifoPath); err != nil {
+				return err
+			}
+			m.Cfg.MetricsFifo = newMetricFifoPath
+
+			return nil
+		},
+	}
+}
+
+// StubDrivesOptions will return a set of options used to create a new stub
+// drive handler.
+func (j runcJailer) StubDrivesOptions() []stubDrivesOpt {
+	return []stubDrivesOpt{
+		func(drives []models.Drive) error {
+			for _, drive := range drives {
+				path := firecracker.StringValue(drive.PathOnHost)
+				if err := os.Chown(path, int(j.uid), int(j.gid)); err != nil {
+					return err
+				}
+			}
+			return nil
+		},
+	}
+}
+
+// ExposeDeviceToJail will inspect the given file, srcDevicePath, and based on the
+// file type, proper handling will occur to ensure that the file is visible in
+// the jail. For block devices we will use mknod to create the device and then
+// set the correct permissions to ensure visibility in the jail.
+func (j runcJailer) ExposeDeviceToJail(srcDevicePath string) error {
+	uid := j.uid
+	gid := j.gid
+
+	stat := syscall.Stat_t{}
+	if err := syscall.Stat(srcDevicePath, &stat); err != nil {
+		return err
+	}
+
+	// Checks file type using S_IFMT which is the bit mask for the file type.
+	// Here we only care about block devices, ie S_IFBLK.  If it is a block type
+	// we will manually call mknod and create that device.
+	if (stat.Mode & syscall.S_IFMT) == syscall.S_IFBLK {
+		path := filepath.Join(j.RootPath(), "dev")
+		dst := filepath.Join(path, filepath.Base(srcDevicePath))
+		if err := exposeBlockDeviceToJail(dst, int(stat.Rdev), int(uid), int(gid)); err != nil {
+			return err
+		}
+	} else {
+		return fmt.Errorf("unsupported mode: %v", stat.Mode)
+	}
+
+	return nil
+}
+
+// exposeBlockDeviceToJail will call mknod on the block device to ensure
+// visibility of the device
+func exposeBlockDeviceToJail(dst string, rdev, uid, gid int) error {
+	if err := syscall.Mknod(dst, syscall.S_IFBLK, rdev); err != nil {
+		return err
+	}
+
+	if err := os.Chmod(dst, 0600); err != nil {
+		return err
+	}
+
+	if err := os.Chown(dst, uid, gid); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func copyFile(src, dst string, mode os.FileMode) error {
+	srcFile, err := os.Open(src)
+	if err != nil {
+		return errors.Wrapf(err, "failed to open %v", src)
+	}
+	defer srcFile.Close()
+
+	dstFile, err := os.OpenFile(dst, os.O_CREATE|os.O_WRONLY|os.O_EXCL, mode)
+	if err != nil {
+		return errors.Wrapf(err, "failed to open %v", dstFile)
+	}
+	defer dstFile.Close()
+
+	_, err = io.Copy(dstFile, srcFile)
+	if err != nil {
+		return errors.Wrap(err, "failed to copy to destination")
+	}
+	return nil
+}
+
+func (j runcJailer) jailerCommand(containerName string) *exec.Cmd {
+	cmd := exec.Command(j.runcBinaryPath, "run", containerName)
+	cmd.Dir = j.OCIBundlePath()
+	return cmd
```

Comment:
This should be `exec.CommandContext` and provided the `shimCtx` so that runc/Firecracker shuts down if the shim shuts down.

Reply:
Good point! Fixed.